### PR TITLE
Week two task

### PR DESCRIPTION
**Pull request by onfranciis**
- [x] Create credit
- [x] Create admin
- [x] Reorganise thead_local
- [x] Create setting

### CODE REVIEW

File: src/syntax_backend/src/schema/setting.rs [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+use candid::{CandidType, Decode, Encode};
+use ic_stable_structures::storable::{Bound, Storable};
+use serde::{Deserialize, Serialize};
+use std::borrow::Cow;
+
+const MAX_VALUE_SIZE: u32 = 1000000;
+pub const SETTING_KEY: &str = "Settings";
+
+#[derive(CandidType, Serialize, Deserialize, Debug, Clone)]
+pub struct Setting {
```
Comment (ogbanugot) [inline src/syntax_backend/src/schema/setting.rs:11]: Hi @onfranciis I don't understand the purpose of this schema. What's the idea here please, I don't recall this being part of what we discussed.

File: src/syntax_backend/src/service/credit.rs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            .map(|(_, value)| value)
+            .collect::<Vec<Credit>>()
+    });
+
+    res
+}
+
+#[ic_cdk::update]
+async fn add_single_credit_plan(plan: String, credit: u16, pwd: String) -> CreditResponse {
+    match SETTING_MAP.with(|map| map.borrow().get(&SETTING_KEY.to_string())) {
```
Comment (ogbanugot) [inline src/syntax_backend/src/service/credit.rs:28]: If SETTING_KEY is supposed to be an API_KEY then perhaps I see what you were going for, but then I already mentioned that we won't be doing it this way. We should instead check that the identity calling these admin endpoints belongs to an AdminMap which should contain theidentities that are admins.

File: src/syntax_backend/src/service/cv.rs [hunk 3]
```diff
@@ -1,11 +1,10 @@
...
 use crate::service::ai;
-use crate::storage::cv::CV_STORAGE_MAP;
+use crate::storage::thread_local::CV_STORAGE_MAP;
 use crate::QUOTA_ERROR;
 use crate::{storage, MONTHLY_TRIAL_ERROR};
 
-use super::util::has_used_up_trials;
-
 #[ic_cdk::update]
 async fn analyze_cv(principal: String, request: CVUserInput) -> CVResponse {
```
Comment (ogbanugot) [inline src/syntax_backend/src/service/cv.rs:11]: How come there are no checks for credit usage in the CV service?  Every time a user tries to analyze a CV, we should check the credit usage and deduct. Why is the logic for that not here on the CV service endpoints?

File: src/syntax_backend/src/schema/admin.rs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+use serde::{Deserialize, Serialize};
+use std::borrow::Cow;
+
+const MAX_VALUE_SIZE: u32 = 1000000;
+pub const NO_ADMIN_FOUND: &str = "No admin with this ID was found!";
+pub const ID_GENERATION_FAILED: &str = "There was an error while generating the ID for this admin!";
+pub const INVALID_AUTH: &str = "Invalid auth password!";
+
+#[derive(Serialize, Deserialize, CandidType, Debug, Clone)]
+pub enum AdminPlan {
```
Comment (ogbanugot) [inline src/syntax_backend/src/schema/admin.rs]: Change to `AdminRole`

**Review by ogbanugot — CHANGES_REQUESTED** [review]

**Comment by ogbanugot** [comment]
Hey @onfranciis this doesn't really implement most of what we discussed. These are the key points we discussed:

1. A service endpoint to tie a user to a subscription tier. Here we are assigning a number of credits (based on the tier type) to this user.

Create a service ‘Credits’
/add_credits (registering credits for a user based on the tier, check if they have credits already and increment based on the tier.)
/get_tiers => returns the available tiers (and credits)

2. Admin Service 
/set_tiers => set the number of credits per $ for each tier.

3. We need to handle managing these credits for each request. We need to check that the user still has enough credits to make a request. And we need to decrement their credit count each time a (CV Analyze) request is made. 

4. We need to remove the identity from service requests and make use of the ICP caller method to fetch the identity (see here https://github.com/Tech-Shark/syntax/blob/f446a67610dcbbccd34e852482ebb61664a16085/src/syntax_backend/src/lib.rs#L13)

File: src/syntax_backend/src/schema/setting.rs (hunk 1 above)
Comment (onfranciis) [reply src/syntax_backend/src/schema/setting.rs:11]: This struct will be used in holding the config/settings related to various parts of this app. This data can be set and updated by only an admin. For example, the `max_freemium_users` is used to know how many people we want to accomodate on the free tier in the app. An admin might decide to reduce or increase the limit.

**Review by ogbanugot — APPROVED** [review]

**Pull request merged**